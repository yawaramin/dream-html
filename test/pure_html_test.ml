(* Copyright 2023 Yawar Amin

   This file is part of dream-html.

   dream-html is free software: you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the Free
   Software Foundation, either version 3 of the License, or (at your option) any
   later version.

   dream-html is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
   details.

   You should have received a copy of the GNU General Public License along with
   dream-html. If not, see <https://www.gnu.org/licenses/>. *)

open Pure_html

let test_html msg node = Format.printf "\n\n‚úîÔ∏é %s\n%a\n" msg pp node

let test_xml ?header msg node =
  Format.printf "\n\nüîé %s\n%a\n" msg (pp_xml ?header) node

let () =
  test_html "HTML"
  @@
  let open HTML in
  let greet nm = p [id "greet-%s" nm] [txt "Hello, %s!" nm] in
  html
    [lang "en"]
    [ head []
        [ title [] "Dream_html Test";
          link [rel "preload"; as_ "style"; href "/app.css"] ];
      body
        [id "test-content"]
        [ main
            [spellcheck true; colspan 1]
            [ article
                [id "article-1"; class_ "story"]
                [ p
                    [Hx.get "/p1?a b"; Hx.target "closest article > p"]
                    [txt "Test para 1."];
                  p [] [txt "Test para 2."];
                  a [href "/a?b=cd:efg/hij"] [txt "cd:efg/hij"];
                  a [href "/foo?a=1&b=2 3&c=4<5&d=6>5"] [txt "Test"];
                  a [href "/üòâ"; Hx.headers {|{"foo": 1}|}] [txt "wink"];
                  MathML.math [style_ ""] [] ];
              input
                [ type_ "text";
                  autocomplete `name;
                  title_ "<script>alert('oops')";
                  onblur "if (1 > 0) alert(this.value)" ];
              null
                [ comment "oops --><script>alert('lol')</script>";
                  dialog [open_; title_ {|"hello"|}] [div [] []];
                  template [id "idtmpl"] [p [] [txt "Template"]];
                  div [translate `no] [p [translate `yes] []];
                  textarea
                    [ required;
                      Hx.trigger "keyup[target.value.trim() != '']";
                      autocapitalize `words ]
                    "'super'";
                  hr [(if true then class_ "super" else null_)];
                  greet "Bob" ] ] ] ]

let () =
  test_html "HTML pretty print corner cases"
  @@
  let open HTML in
  (* test empty tags don't break line *)
  div [] [p [translate `yes] []]

let () =
  let open HTML in
  test_html "Concat HTML - empty list" (concat (txt ", ") []);
  test_html "Concat HTML - singleton list"
    (concat (txt ", ") [a [href "/"] [txt "Home"]]);
  test_html "Concat HTML - list"
    (concat (txt ", ") [a [href "/"] [txt "Home"]; a [href "/a"] [txt "a"]])

let () =
  let open HTML in
  try ignore (div [id "foo"] [] +@ id "bar")
  with Invalid_argument _ ->
    Format.printf "\n\n‚úîÔ∏é test raise on adding duplicate attribute\n"

let () =
  let open HTML in
  test_html "add class attribute to existing"
    (div [class_ "foo"] [] +@ class_ "bar")

let () =
  let open HTML in
  test_html "add new attribute" (div [id "foo"] [] +@ class_ "bar")

let () =
  let node =
    let open HTML in
    div [class_ "a b c"] [p [class_ "d e f"] []; p [class_ "g h i"] []]
  in
  node
  |> fold
       ~tag:(fun _name attrs classes ->
         match List.find_opt (fun (n, _) -> n = "class") attrs with
         | Some (_name, c) -> classes ^ " " ^ c
         | None -> classes)
       ~txt:(fun _string c -> c)
       ~comment:(fun _string c -> c)
       ""
  |> Printf.printf "\n\n‚úîÔ∏é fold node: %s\n"

let () =
  test_xml ~header:true "SVG"
  @@
  let open SVG in
  svg
    [ xmlns;
      fill "none";
      viewbox ~min_x:0 ~min_y:0 ~width:24 ~height:24;
      stroke_width "1.5";
      stroke "currentColor";
      HTML.class_ "w-6 h-6" ]
    [ path
        [ stroke_linecap `round;
          stroke_linejoin `round;
          d "M4.5 10.5L12 3m0 0l7.5 7.5M12 3v18" ]
        [] ]

let () =
  test_xml "MathML"
  @@
  let open HTML in
  let open MathML in
  p []
    [ math
        [style_ ""]
        [ mtable []
            [ mtr []
                [ mtd []
                    [ mrow []
                        [ mrow []
                            [ mrow []
                                [ mrow []
                                    [ mi [] [txt "a"];
                                      mo [] [txt "‚Å¢"];
                                      msup [] [mi [] [txt "x"]; mn [] [txt "2"]]
                                    ];
                                  mo [] [txt "+"];
                                  mi [] [txt "b"];
                                  mo [] [txt "‚Å¢"];
                                  mi [] [txt "x"] ];
                              mo [] [txt "+"];
                              mi [] [txt "c"] ] ] ];
                  mtd [] [mo [] [txt "="]];
                  mtd [] [mn [] [txt "0"]] ];
              mtr []
                [ mtd []
                    [ mrow []
                        [ mrow []
                            [ mi [] [txt "a"];
                              mo [] [txt "‚Å¢"];
                              msup [] [mi [] [txt "x"]; mn [] [txt "2"]] ];
                          mo [] [txt "+"];
                          mi [] [txt "b"];
                          mo [] [txt "‚Å¢"];
                          mi [] [txt "x"] ] ];
                  mtd [] [mo [] [txt "="]];
                  mtd [] [mo [] [txt "‚àí"]; mi [] [txt "c"]] ];
              mtr []
                [ mtd []
                    [ mrow []
                        [ mrow [] [msup [] [mi [] [txt "x"]; mn [] [txt "2"]]];
                          mo [] [txt "+"];
                          mfrac [] [mi [] [txt "b"]; mi [] [txt "a"]];
                          mo [] [txt "‚Å§"];
                          mi [] [txt "x"] ] ];
                  mtd [] [mo [] [txt "="]];
                  mtd []
                    [ mfrac []
                        [ mrow [] [mo [] [txt "‚àí"]; mi [] [txt "c"]];
                          mi [] [txt "a"] ] ];
                  mtd []
                    [ mrow []
                        [ mtext
                            [style_ "color: red; font-size: 10pt;"]
                            [txt "Divide out leading coefficient."] ] ] ];
              mtr []
                [ mtd []
                    [ mrow []
                        [ mrow []
                            [ mrow []
                                [msup [] [mi [] [txt "x"]; mn [] [txt "2"]]];
                              mo [] [txt "+"];
                              mfrac []
                                [mrow [] [mi [] [txt "b"]]; mi [] [txt "a"]];
                              mo [] [txt "‚Å§"];
                              mi [] [txt "x"];
                              mo [] [txt "+"];
                              msup []
                                [ mrow []
                                    [ mo [] [txt "("];
                                      mfrac []
                                        [ mrow [] [mi [] [txt "b"]];
                                          mrow []
                                            [mn [] [txt "2"]; mi [] [txt "a"]]
                                        ];
                                      mo [] [txt ")"] ];
                                  mn [] [txt "2"] ] ] ] ];
                  mtd [] [mo [] [txt "="]];
                  mtd []
                    [ mrow []
                        [ mfrac []
                            [ mrow []
                                [ mo [] [txt "‚àí"];
                                  mi [] [txt "c"];
                                  mo [] [txt "("];
                                  mn [] [txt "4"];
                                  mi [] [txt "a"];
                                  mo [] [txt ")"] ];
                              mrow []
                                [ mi [] [txt "a"];
                                  mo [] [txt "("];
                                  mn [] [txt "4"];
                                  mi [] [txt "a"];
                                  mo [] [txt ")"] ] ];
                          mo [] [txt "+"];
                          mfrac []
                            [ mrow []
                                [msup [] [mi [] [txt "b"]; mn [] [txt "2"]]];
                              mrow []
                                [ mn [] [txt "4"];
                                  msup [] [mi [] [txt "a"]; mn [] [txt "2"]] ]
                            ] ] ];
                  mtd []
                    [ mrow []
                        [ mtext
                            [style_ "color: red; font-size: 10pt;"]
                            [txt "Complete the square."] ] ] ];
              mtr []
                [ mtd []
                    [ mrow []
                        [ mrow []
                            [ mo [] [txt "("];
                              mi [] [txt "x"];
                              mo [] [txt "+"];
                              mfrac []
                                [ mrow [] [mi [] [txt "b"]];
                                  mrow [] [mn [] [txt "2"]; mi [] [txt "a"]] ];
                              mo [] [txt ")"];
                              mo [] [txt "("];
                              mi [] [txt "x"];
                              mo [] [txt "+"];
                              mfrac []
                                [ mrow [] [mi [] [txt "b"]];
                                  mrow [] [mn [] [txt "2"]; mi [] [txt "a"]] ];
                              mo [] [txt ")"] ] ] ];
                  mtd [] [mo [] [txt "="]];
                  mtd []
                    [ mfrac []
                        [ mrow []
                            [ msup [] [mi [] [txt "b"]; mn [] [txt "2"]];
                              mo [] [txt "‚àí"];
                              mn [] [txt "4"];
                              mi [] [txt "a"];
                              mi [] [txt "c"] ];
                          mrow []
                            [ mn [] [txt "4"];
                              msup [] [mi [] [txt "a"]; mn [] [txt "2"]] ] ] ];
                  mtd []
                    [ mrow []
                        [ mtext
                            [style_ "color: red; font-size: 10pt;"]
                            [txt "Discriminant revealed."] ] ] ];
              mtr []
                [ mtd []
                    [ mrow []
                        [ mrow []
                            [ msup []
                                [ mrow []
                                    [ mo [] [txt "("];
                                      mi [] [txt "x"];
                                      mo [] [txt "+"];
                                      mfrac []
                                        [ mrow [] [mi [] [txt "b"]];
                                          mrow []
                                            [mn [] [txt "2"]; mi [] [txt "a"]]
                                        ];
                                      mo [] [txt ")"] ];
                                  mn [] [txt "2"] ] ] ] ];
                  mtd [] [mo [] [txt "="]];
                  mtd []
                    [ mfrac []
                        [ mrow []
                            [ msup [] [mi [] [txt "b"]; mn [] [txt "2"]];
                              mo [] [txt "‚àí"];
                              mn [] [txt "4"];
                              mi [] [txt "a"];
                              mi [] [txt "c"] ];
                          mrow []
                            [ mn [] [txt "4"];
                              msup [] [mi [] [txt "a"]; mn [] [txt "2"]] ] ] ];
                  mtd []
                    [mrow [] [mtext [style_ "color: red; font-size: 10pt;"] []]]
                ];
              mtr []
                [ mtd []
                    [ mrow []
                        [ mrow []
                            [ mrow []
                                [ mi [] [txt "x"];
                                  mo [] [txt "+"];
                                  mfrac []
                                    [ mrow [] [mi [] [txt "b"]];
                                      mrow [] [mn [] [txt "2"]; mi [] [txt "a"]]
                                    ] ] ] ] ];
                  mtd [] [mo [] [txt "="]];
                  mtd []
                    [ msqrt []
                        [ mfrac []
                            [ mrow []
                                [ msup [] [mi [] [txt "b"]; mn [] [txt "2"]];
                                  mo [] [txt "‚àí"];
                                  mn [] [txt "4"];
                                  mi [] [txt "a"];
                                  mi [] [txt "c"] ];
                              mrow []
                                [ mn [] [txt "4"];
                                  msup [] [mi [] [txt "a"]; mn [] [txt "2"]] ]
                            ] ] ];
                  mtd []
                    [mrow [] [mtext [style_ "color: red; font-size: 10pt;"] []]]
                ];
              mtr []
                [ mtd [] [mi [] [txt "x"]];
                  mtd [] [mo [] [txt "="]];
                  mtd []
                    [ mfrac []
                        [ mrow [] [mo [] [txt "‚àí"]; mi [] [txt "b"]];
                          mrow [] [mn [] [txt "2"]; mi [] [txt "a"]] ];
                      mo [] [txt "¬±"];
                      mrow [] [mo [] [txt "{"]; mi [] [txt "C"]; mo [] [txt "}"]];
                      msqrt []
                        [ mfrac []
                            [ mrow []
                                [ msup [] [mi [] [txt "b"]; mn [] [txt "2"]];
                                  mo [] [txt "‚àí"];
                                  mn [] [txt "4"];
                                  mi [] [txt "a"];
                                  mi [] [txt "c"] ];
                              mrow []
                                [ mn [] [txt "4"];
                                  msup [] [mi [] [txt "a"]; mn [] [txt "2"]] ]
                            ] ] ];
                  mtd []
                    [ mrow []
                        [ mtext
                            [style_ "color: red; font-size: 10pt;"]
                            [txt "There's the vertex formula."] ] ] ];
              mtr []
                [ mtd [] [mi [] [txt "x"]];
                  mtd [] [mo [] [txt "="]];
                  mtd []
                    [ mfrac []
                        [ mrow []
                            [ mo [] [txt "‚àí"];
                              mi [] [txt "b"];
                              mo [] [txt "¬±"];
                              mrow []
                                [ mo [] [txt "{"];
                                  mi [] [txt "C"];
                                  mo [] [txt "}"] ];
                              msqrt []
                                [ msup [] [mi [] [txt "b"]; mn [] [txt "2"]];
                                  mo [] [txt "‚àí"];
                                  mn [] [txt "4"];
                                  mi [] [txt "a"];
                                  mi [] [txt "c"] ] ];
                          mrow [] [mn [] [txt "2"]; mi [] [txt "a"]] ] ];
                  mtd []
                    [mrow [] [mtext [style_ "color: red; font-size: 10pt;"] []]]
                ] ] ] ]

let () =
  test_xml ~header:true "RSS"
  @@
  let open Pure_html in
  let open RSS in
  rss [Atom.xmlns; version_2]
    [ channel []
        [ title [] "NASA Space Station News";
          link [] "http://www.nasa.gov/";
          description []
            "A RSS news feed containing the latest NASA press releases on the International Space Station.";
          language [] "en-us";
          pub_date [] "Tue, 10 Jun 2003 04:00:00 GMT";
          last_build_date [] "Fri, 21 Jul 2023 09:04 EDT";
          docs [] "https://www.rssboard.org/rss-specification";
          generator [] "pure-html";
          managing_editor [] "neil.armstrong@example.com (Neil Armstrong)";
          web_master [] "sally.ride@example.com (Sally Ride)";
          Atom.link
            HTML.
              [ href "https://www.rssboard.org/files/sample-rss-2.xml";
                rel "self";
                type_ "application/rss+xml" ]
            [];
          item []
            [ title []
                "Louisiana Students to Hear from NASA Astronauts Aboard Space Station";
              link []
                "http://www.nasa.gov/press-release/louisiana-students-to-hear-from-nasa-astronauts-aboard-space-station";
              description []
                "As part of the state's first Earth-to-space call, students from Louisiana will have an opportunity soon to hear from NASA astronauts aboard the International Space Station.";
              pub_date [] "Fri, 21 Jul 2023 09:04 EDT";
              guid []
                "http://www.nasa.gov/press-release/louisiana-students-to-hear-from-nasa-astronauts-aboard-space-station"
            ] ] ]
