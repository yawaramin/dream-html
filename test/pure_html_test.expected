

‚úîÔ∏é HTML
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Dream_html Test</title>
    <link rel="preload" as="style" href="/app.css">
  </head>
  <body id="test-content">
    <main spellcheck="true" colspan="1">
      <article id="article-1" class="story">
        <p data-hx-get="/p1?a%20b" data-hx-target="closest article > p">Test para 1.</p>
        <p>Test para 2.</p>
        <a href="/a?b=cd:efg/hij">cd:efg/hij</a>
        <a href="/foo?a=1&amp;b=2%203&amp;c=4%3C5&amp;d=6%3E5">Test</a>
        <a href="/%F0%9F%98%89" data-hx-headers='{"foo": 1}'>wink</a>
        <math style=""></math>
      </article>
      <input type="text" autocomplete="name" onblur="if (1 > 0) alert(this.value)">
      <!-- oops --&gt;&lt;script&gt;alert('lol')&lt;/script&gt; -->
      <dialog open title="&quot;hello&quot;">
        <div></div>
      </dialog>
      <template id="idtmpl">
        <p>Template</p>
      </template>
      <div translate="no">
        <p translate="yes"></p>
      </div>
      <textarea required data-hx-trigger="keyup[target.value.trim() != '']" autocapitalize="words">'super'</textarea>
      <hr class="super">
      <p id="greet-Bob">Hello, Bob!</p>
    </main>
  </body>
</html>



‚úîÔ∏é HTML pretty print corner cases
<div>
  <p translate="yes"></p>
</div>



‚úîÔ∏é Concat HTML - empty list



‚úîÔ∏é Concat HTML - singleton list
<a href="/">Home</a>



‚úîÔ∏é Concat HTML - list
<a href="/">Home</a>, <a href="/a">a</a>


‚úîÔ∏é test raise on adding duplicate attribute


‚úîÔ∏é add class attribute to existing
<div class="foo bar"></div>



‚úîÔ∏é add new attribute
<div class="bar" id="foo"></div>



üîé SVG
<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewbox="0 0 24 24" stroke-width="1.5" stroke="currentColor" class="w-6 h-6">
  <path stroke-linecap="round" stroke-linejoin="round" d="M4.5 10.5L12 3m0 0l7.5 7.5M12 3v18" />
</svg>



üîé MathML
<p>
  <math style="">
    <mtable>
      <mtr>
        <mtd>
          <mrow>
            <mrow>
              <mrow>
                <mrow>
                  <mi>a</mi>
                  <mo>‚Å¢</mo>
                  <msup>
                    <mi>x</mi>
                    <mn>2</mn>
                  </msup>
                </mrow>
                <mo>+</mo>
                <mi>b</mi>
                <mo>‚Å¢</mo>
                <mi>x</mi>
              </mrow>
              <mo>+</mo>
              <mi>c</mi>
            </mrow>
          </mrow>
        </mtd>
        <mtd>
          <mo>=</mo>
        </mtd>
        <mtd>
          <mn>0</mn>
        </mtd>
      </mtr>
      <mtr>
        <mtd>
          <mrow>
            <mrow>
              <mi>a</mi>
              <mo>‚Å¢</mo>
              <msup>
                <mi>x</mi>
                <mn>2</mn>
              </msup>
            </mrow>
            <mo>+</mo>
            <mi>b</mi>
            <mo>‚Å¢</mo>
            <mi>x</mi>
          </mrow>
        </mtd>
        <mtd>
          <mo>=</mo>
        </mtd>
        <mtd>
          <mo>‚àí</mo>
          <mi>c</mi>
        </mtd>
      </mtr>
      <mtr>
        <mtd>
          <mrow>
            <mrow>
              <msup>
                <mi>x</mi>
                <mn>2</mn>
              </msup>
            </mrow>
            <mo>+</mo>
            <mfrac>
              <mi>b</mi>
              <mi>a</mi>
            </mfrac>
            <mo>‚Å§</mo>
            <mi>x</mi>
          </mrow>
        </mtd>
        <mtd>
          <mo>=</mo>
        </mtd>
        <mtd>
          <mfrac>
            <mrow>
              <mo>‚àí</mo>
              <mi>c</mi>
            </mrow>
            <mi>a</mi>
          </mfrac>
        </mtd>
        <mtd>
          <mrow>
            <mtext style="color: red; font-size: 10pt;">Divide out leading coefficient.</mtext>
          </mrow>
        </mtd>
      </mtr>
      <mtr>
        <mtd>
          <mrow>
            <mrow>
              <mrow>
                <msup>
                  <mi>x</mi>
                  <mn>2</mn>
                </msup>
              </mrow>
              <mo>+</mo>
              <mfrac>
                <mrow>
                  <mi>b</mi>
                </mrow>
                <mi>a</mi>
              </mfrac>
              <mo>‚Å§</mo>
              <mi>x</mi>
              <mo>+</mo>
              <msup>
                <mrow>
                  <mo>(</mo>
                  <mfrac>
                    <mrow>
                      <mi>b</mi>
                    </mrow>
                    <mrow>
                      <mn>2</mn>
                      <mi>a</mi>
                    </mrow>
                  </mfrac>
                  <mo>)</mo>
                </mrow>
                <mn>2</mn>
              </msup>
            </mrow>
          </mrow>
        </mtd>
        <mtd>
          <mo>=</mo>
        </mtd>
        <mtd>
          <mrow>
            <mfrac>
              <mrow>
                <mo>‚àí</mo>
                <mi>c</mi>
                <mo>(</mo>
                <mn>4</mn>
                <mi>a</mi>
                <mo>)</mo>
              </mrow>
              <mrow>
                <mi>a</mi>
                <mo>(</mo>
                <mn>4</mn>
                <mi>a</mi>
                <mo>)</mo>
              </mrow>
            </mfrac>
            <mo>+</mo>
            <mfrac>
              <mrow>
                <msup>
                  <mi>b</mi>
                  <mn>2</mn>
                </msup>
              </mrow>
              <mrow>
                <mn>4</mn>
                <msup>
                  <mi>a</mi>
                  <mn>2</mn>
                </msup>
              </mrow>
            </mfrac>
          </mrow>
        </mtd>
        <mtd>
          <mrow>
            <mtext style="color: red; font-size: 10pt;">Complete the square.</mtext>
          </mrow>
        </mtd>
      </mtr>
      <mtr>
        <mtd>
          <mrow>
            <mrow>
              <mo>(</mo>
              <mi>x</mi>
              <mo>+</mo>
              <mfrac>
                <mrow>
                  <mi>b</mi>
                </mrow>
                <mrow>
                  <mn>2</mn>
                  <mi>a</mi>
                </mrow>
              </mfrac>
              <mo>)</mo>
              <mo>(</mo>
              <mi>x</mi>
              <mo>+</mo>
              <mfrac>
                <mrow>
                  <mi>b</mi>
                </mrow>
                <mrow>
                  <mn>2</mn>
                  <mi>a</mi>
                </mrow>
              </mfrac>
              <mo>)</mo>
            </mrow>
          </mrow>
        </mtd>
        <mtd>
          <mo>=</mo>
        </mtd>
        <mtd>
          <mfrac>
            <mrow>
              <msup>
                <mi>b</mi>
                <mn>2</mn>
              </msup>
              <mo>‚àí</mo>
              <mn>4</mn>
              <mi>a</mi>
              <mi>c</mi>
            </mrow>
            <mrow>
              <mn>4</mn>
              <msup>
                <mi>a</mi>
                <mn>2</mn>
              </msup>
            </mrow>
          </mfrac>
        </mtd>
        <mtd>
          <mrow>
            <mtext style="color: red; font-size: 10pt;">Discriminant revealed.</mtext>
          </mrow>
        </mtd>
      </mtr>
      <mtr>
        <mtd>
          <mrow>
            <mrow>
              <msup>
                <mrow>
                  <mo>(</mo>
                  <mi>x</mi>
                  <mo>+</mo>
                  <mfrac>
                    <mrow>
                      <mi>b</mi>
                    </mrow>
                    <mrow>
                      <mn>2</mn>
                      <mi>a</mi>
                    </mrow>
                  </mfrac>
                  <mo>)</mo>
                </mrow>
                <mn>2</mn>
              </msup>
            </mrow>
          </mrow>
        </mtd>
        <mtd>
          <mo>=</mo>
        </mtd>
        <mtd>
          <mfrac>
            <mrow>
              <msup>
                <mi>b</mi>
                <mn>2</mn>
              </msup>
              <mo>‚àí</mo>
              <mn>4</mn>
              <mi>a</mi>
              <mi>c</mi>
            </mrow>
            <mrow>
              <mn>4</mn>
              <msup>
                <mi>a</mi>
                <mn>2</mn>
              </msup>
            </mrow>
          </mfrac>
        </mtd>
        <mtd>
          <mrow>
            <mtext style="color: red; font-size: 10pt;" />
          </mrow>
        </mtd>
      </mtr>
      <mtr>
        <mtd>
          <mrow>
            <mrow>
              <mrow>
                <mi>x</mi>
                <mo>+</mo>
                <mfrac>
                  <mrow>
                    <mi>b</mi>
                  </mrow>
                  <mrow>
                    <mn>2</mn>
                    <mi>a</mi>
                  </mrow>
                </mfrac>
              </mrow>
            </mrow>
          </mrow>
        </mtd>
        <mtd>
          <mo>=</mo>
        </mtd>
        <mtd>
          <msqrt>
            <mfrac>
              <mrow>
                <msup>
                  <mi>b</mi>
                  <mn>2</mn>
                </msup>
                <mo>‚àí</mo>
                <mn>4</mn>
                <mi>a</mi>
                <mi>c</mi>
              </mrow>
              <mrow>
                <mn>4</mn>
                <msup>
                  <mi>a</mi>
                  <mn>2</mn>
                </msup>
              </mrow>
            </mfrac>
          </msqrt>
        </mtd>
        <mtd>
          <mrow>
            <mtext style="color: red; font-size: 10pt;" />
          </mrow>
        </mtd>
      </mtr>
      <mtr>
        <mtd>
          <mi>x</mi>
        </mtd>
        <mtd>
          <mo>=</mo>
        </mtd>
        <mtd>
          <mfrac>
            <mrow>
              <mo>‚àí</mo>
              <mi>b</mi>
            </mrow>
            <mrow>
              <mn>2</mn>
              <mi>a</mi>
            </mrow>
          </mfrac>
          <mo>¬±</mo>
          <mrow>
            <mo>{</mo>
            <mi>C</mi>
            <mo>}</mo>
          </mrow>
          <msqrt>
            <mfrac>
              <mrow>
                <msup>
                  <mi>b</mi>
                  <mn>2</mn>
                </msup>
                <mo>‚àí</mo>
                <mn>4</mn>
                <mi>a</mi>
                <mi>c</mi>
              </mrow>
              <mrow>
                <mn>4</mn>
                <msup>
                  <mi>a</mi>
                  <mn>2</mn>
                </msup>
              </mrow>
            </mfrac>
          </msqrt>
        </mtd>
        <mtd>
          <mrow>
            <mtext style="color: red; font-size: 10pt;">There's the vertex formula.</mtext>
          </mrow>
        </mtd>
      </mtr>
      <mtr>
        <mtd>
          <mi>x</mi>
        </mtd>
        <mtd>
          <mo>=</mo>
        </mtd>
        <mtd>
          <mfrac>
            <mrow>
              <mo>‚àí</mo>
              <mi>b</mi>
              <mo>¬±</mo>
              <mrow>
                <mo>{</mo>
                <mi>C</mi>
                <mo>}</mo>
              </mrow>
              <msqrt>
                <msup>
                  <mi>b</mi>
                  <mn>2</mn>
                </msup>
                <mo>‚àí</mo>
                <mn>4</mn>
                <mi>a</mi>
                <mi>c</mi>
              </msqrt>
            </mrow>
            <mrow>
              <mn>2</mn>
              <mi>a</mi>
            </mrow>
          </mfrac>
        </mtd>
        <mtd>
          <mrow>
            <mtext style="color: red; font-size: 10pt;" />
          </mrow>
        </mtd>
      </mtr>
    </mtable>
  </math>
</p>



üîé RSS
<?xml version="1.0" encoding="UTF-8"?>
<rss xmlns:atom="http://www.w3.org/2005/Atom" version="2.0">
  <channel>
    <title>NASA Space Station News</title>
    <link>http://www.nasa.gov/</link>
    <description>A RSS news feed containing the latest NASA press releases on the International Space Station.</description>
    <language>en-us</language>
    <pubDate>Tue, 10 Jun 2003 04:00:00 GMT</pubDate>
    <lastBuildDate>Fri, 21 Jul 2023 09:04 EDT</lastBuildDate>
    <docs>https://www.rssboard.org/rss-specification</docs>
    <generator>pure-html</generator>
    <managingEditor>neil.armstrong@example.com (Neil Armstrong)</managingEditor>
    <webMaster>sally.ride@example.com (Sally Ride)</webMaster>
    <atom:link href="https://www.rssboard.org/files/sample-rss-2.xml" rel="self" type="application/rss+xml" />
    <item>
      <title>Louisiana Students to Hear from NASA Astronauts Aboard Space Station</title>
      <link>http://www.nasa.gov/press-release/louisiana-students-to-hear-from-nasa-astronauts-aboard-space-station</link>
      <description>As part of the state's first Earth-to-space call, students from Louisiana will have an opportunity soon to hear from NASA astronauts aboard the International Space Station.</description>
      <pubDate>Fri, 21 Jul 2023 09:04 EDT</pubDate>
      <guid>http://www.nasa.gov/press-release/louisiana-students-to-hear-from-nasa-astronauts-aboard-space-station</guid>
    </item>
  </channel>
</rss>

